configfile: "./config/config.yaml"

# Rules to include
include: "./rules/fastqc.smk"
include: "./rules/merge.smk"


# Define the directory containing the FASTQ files
sample_dir = config["sample_dir"]

# Define the output directory
output_dir = config["output_dir"]

# Define patterns to match specific files
samples = glob_wildcards(f"{sample_dir}/{{sample}}.fastq.gz").sample


rule all:
    input:
        expand("{output_dir}/qc/{sample}_fastqc.html", 
               sample=samples, 
               output_dir=output_dir)

rule trim_adapters:
    input:
        fq="data/{sample}.fastq.gz"
    output:
        trimmed="results/trimmed/{sample}_trimmed.fastq.gz"
    log:
        "logs/trim/{sample}.log"
    shell:
        "cutadapt -a AGATCGGAAGAG -o {output.trimmed} {input.fq} > {log} 2>&1"

rule align:
    input:
        fq="results/trimmed/{sample}_trimmed.fastq.gz",
        genome=config["genome"]
    output:
        bam="results/aligned/{sample}.bam"
    log:
        "logs/align/{sample}.log"
    shell:
        "bowtie2 -x {input.genome} -U {input.fq} | samtools view -b > {output.bam} 2> {log}"